File: python3/Decode.py
class Decode:
	def decode(self,s):
		"""
		:type s: string
		:rtype: List[string]
		"""
		result = []
		# we are going to use an empty array and join it together, acting in place of StringBuilder
		self.decode_helper(s, 0, [], result)
		return result

	def decode_helper(self, s, i, curr, result):
		"""
		:type s: string
		:type i: int
		:type curr: List[string]
		:type result: List[string]
		:rtype: void
		"""
		if i >= len(s):
			result.append(''.join(curr))
			return

		for c in self.get_next(s,i):
			curr.append(c)
			if ord(c) < ord('A') + 10:
				self.decode_helper(s, i + 1, curr, result)
			else:
				self.decode_helper(s, i + 2, curr, result)
			del curr[-1]

	def get_next(self, s, i):
		"""
		:type s: string
		:type i: int
		:rtype: List[string]
		"""
		to_return = []
		val = int(s[i:i+1])
		to_return.append(chr(ord('A') + val - 1))

		if i < len(s)-1:
			val = int(s[i:i+2])
			if val <= 26:
				to_return.append(chr(ord('A') + val - 1))
		return to_return

test = Decode()
print(test.decode("223"))
print(test.decode("227"))
print(test.decode("4231123"))

File: java/Decode.java
import java.util.List;
import java.util.LinkedList;

public class Decode {  
    public static List<String> decode(String s) {
        List<String> result = new LinkedList<String>();
        decode(s, 0, new StringBuilder(), result);
        return result;
    }
    
    private static void decode(String s, int i, StringBuilder curr, List<String> result) {
        if (i >= s.length()) {
            result.add(curr.toString());
            return;
        }
        
        for (char c : getNext(s, i)) {
            curr.append(c);
            
            if (c < 'A'+10) {
                decode(s, i+1, curr, result);
            } else {
                decode(s, i+2, curr, result);
            }
            curr.setLength(curr.length() - 1);
        }
    }
    
    private static List<Character> getNext(String s, int i) {
        List<Character> toReturn = new LinkedList<Character>();
        int val = Integer.valueOf(s.substring(i, i+1));
        toReturn.add((char) ('A' + val - 1));
        if (i < s.length() - 1) {
            val = Integer.valueOf(s.substring(i, i+2));
            if (val <= 26) toReturn.add((char) ('A' + val - 1));
        }
        return toReturn;
    }
    
    public static void main(String[] args) {
        System.out.println(decode("223"));
        System.out.println(decode("227"));
        System.out.println(decode("4231123"));
    }
}